ocamllex raises Invalid_argument String.sub/Bytes.sub and sometimes segfaults

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

I was trying to modify ocaml-mdx's lexer when suddenly I noticed some SEGV in its testsuite.
Unfortunately I don't have the SEGV reproducer anymore, but I was able to 100% reproduce the following error:
```
Raised at Stdlib.invalid_arg in file "stdlib.ml", line 30, characters 20-45
Called from Stdlib__Bytes.sub_string in file "bytes.ml" (inlined), line 73, characters 44-59
Called from Stdlib__Lexing.sub_lexeme in file "lexing.ml" (inlined), line 196, characters 2-43
Called from Mdx__Lexer_mdx.__ocaml_lex_text_rec in file "lib/lexer_mdx.ml", line 11976, characters 2-78
Called from Mdx__Lexer_mdx.__ocaml_lex_text_rec in file "lib/lexer_mdx.mll", line 59, characters 21-40
Called from Mdx__Lexer_mdx.__ocaml_lex_text_rec in file "lib/lexer_mdx.mll", line 59, characters 21-40
Called from Mdx__Lexer_mdx.__ocaml_lex_text_rec in file "lib/lexer_mdx.mll", line 59, characters 21-40
Called from Mdx__Lexer_mdx.markdown_token in file "lib/lexer_mdx.mll", line 126, characters 11-29
```

Which was introduced by this modification to the lexer:
```
   | ( "<!--" ws* "$MDX" ws* ([^' ' '\n']* as label_cmt) ws* "-->" ws* eol? )?
-      "```" ([^' ' '\n']* as header) ws* ([^'\n']* as legacy_labels) eol
+      "```"  ([^' ' '=' '\n']* as header)? ws* ([^ '{' '\n']* as legacy_labels)
+      ('{' (('#'[^' ' '\n']+)? as _identifier) ws* ('.'([^' ' '}' '\n']+) as _header2)? ws* ([^'}' '\n']* as _attributes) '}')?
+      eol
```

If I remove the `'='` from the first character set then it no longer raises that exception.

In other places where the testsuite runs it segfaults, which should definitely not happen.

E.g. using 'coredumpctl gdb' shows:
```
Core was generated by `/var/home/edwin/.opam/4.14.1/bin/ocamlrun ../../../../../../install/default/bin'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  caml_string_equal (s1=275, s2=139980178532800) at str.c:277
Downloading source file /var/home/edwin/.opam/4.14.1/.opam-switch/build/ocaml-base-compiler.4.14.1/runtime/str.c
                                                                                                                                                 
warning: 277    str.c: No such file or directory
(gdb) bt
#0  caml_string_equal (s1=275, s2=139980178532800) at str.c:277
#1  0x000000000042c8a8 in caml_interprete (prog=0x7f4face00010, prog_size=<optimized out>) at interp.c:936
#2  0x000000000042f55c in caml_main (argv=0x7fff1d4aba48) at startup_byt.c:588
#3  0x000000000040f78c in main (argc=<optimized out>, argv=<optimized out>) at main.c:37

```

Happens on 5.1.1 too:
```
rogram terminated with signal SIGSEGV, Segmentation fault.
#0  caml_string_equal (s1=-1, s2=139720215583240) at runtime/str.c:277
                                                                                                                                                 
warning: 277    runtime/str.c: No such file or directory
(gdb) bt
#0  caml_string_equal (s1=-1, s2=139720215583240) at runtime/str.c:277
#1  0x000000000043a28b in caml_interprete (prog=<optimized out>, prog_size=<optimized out>) at runtime/interp.c:1044
#2  0x000000000043bd39 in caml_main (argv=0x7ffd4c9db8c8) at runtime/startup_byt.c:577
#3  0x0000000000411a0c in main (argc=<optimized out>, argv=<optimized out>) at runtime/main.c:37
```

I've pushed the test code here: https://github.com/realworldocaml/mdx/compare/main...edwintorok:mdx:segv, I haven't yet tried to create a minimal testcase, I'll update this bugreport when I do.

--- a/Changes
+++ b/Changes
@@ -453,6 +453,10 @@ OCaml 5.2.0
 - #12744: ocamltest: run tests in recursive subdirs more eagerly
   (Nick Roberts, review by Nicolás Ojeda Bär)
 
+- #12901, 12908: ocamllex: add overflow checks to prevent generating incorrect
+  lexers; use unsigned numbers in the table encoding when possible.
+  (Vincent Laviron, report by Edwin Török, review by Xavier Leroy)
+
 ### Manual and documentation:
 
 - #12338: clarification of the documentation of process related function in

--- a/lex/output.ml
+++ b/lex/output.ml
@@ -20,6 +20,10 @@ open Lexgen
 open Compact
 open Common
 
+type c_short_type = Short | Unsigned_short
+
+exception Table_overflow
+
 (* To output an array of short ints, encoded as a string *)
 
 let output_byte oc b =
@@ -28,14 +32,26 @@ let output_byte oc b =
   output_char oc (Char.chr(48 + (b / 10) mod 10));
   output_char oc (Char.chr(48 + b mod 10))
 
-let output_array oc v =
+let output_array c_type oc v =
   output_string oc "   \"";
   for i = 0 to Array.length v - 1 do
     output_byte oc (v.(i) land 0xFF);
     output_byte oc ((v.(i) asr 8) land 0xFF);
+    let () =
+      let unsigned_number =
+        match c_type with
+        | Short -> v.(i) + 1 lsl 15
+        | Unsigned_short -> v.(i)
+      in
+      if unsigned_number < 0 || unsigned_number >= 1 lsl 16
+      then raise Table_overflow
+    in
     if i land 7 = 7 then output_string oc "\\\n    "
   done;
   output_string oc "\""
+
+let output_array_s = output_array Short
+let output_array_u = output_array Unsigned_short
 
 let output_byte_array oc v =
   output_string oc "   \"";
@@ -50,21 +66,21 @@ let output_byte_array oc v =
 let output_tables oc tbl =
   output_string oc "let __ocaml_lex_tables = {\n";
 
-  fprintf oc "  Lexing.lex_base =\n%a;\n" output_array tbl.tbl_base;
-  fprintf oc "  Lexing.lex_backtrk =\n%a;\n" output_array tbl.tbl_backtrk;
-  fprintf oc "  Lexing.lex_default =\n%a;\n" output_array tbl.tbl_default;
-  fprintf oc "  Lexing.lex_trans =\n%a;\n" output_array tbl.tbl_trans;
-  fprintf oc "  Lexing.lex_check =\n%a;\n" output_array tbl.tbl_check;
-  fprintf oc "  Lexing.lex_base_code =\n%a;\n" output_array tbl.tbl_base_code;
+  fprintf oc "  Lexing.lex_base =\n%a;\n" output_array_s tbl.tbl_base;
+  fprintf oc "  Lexing.lex_backtrk =\n%a;\n" output_array_s tbl.tbl_backtrk;
+  fprintf oc "  Lexing.lex_default =\n%a;\n" output_array_s tbl.tbl_default;
+  fprintf oc "  Lexing.lex_trans =\n%a;\n" output_array_s tbl.tbl_trans;
+  fprintf oc "  Lexing.lex_check =\n%a;\n" output_array_s tbl.tbl_check;
+  fprintf oc "  Lexing.lex_base_code =\n%a;\n" output_array_u tbl.tbl_base_code;
 
   fprintf oc "  Lexing.lex_backtrk_code =\n%a;\n"
-    output_array tbl.tbl_backtrk_code;
+    output_array_u tbl.tbl_backtrk_code;
   fprintf oc "  Lexing.lex_default_code =\n%a;\n"
-    output_array tbl.tbl_default_code;
+    output_array_u tbl.tbl_default_code;
   fprintf oc "  Lexing.lex_trans_code =\n%a;\n"
-    output_array tbl.tbl_trans_code;
+    output_array_u tbl.tbl_trans_code;
   fprintf oc "  Lexing.lex_check_code =\n%a;\n"
-    output_array tbl.tbl_check_code;
+    output_array_s tbl.tbl_check_code;
   fprintf oc "  Lexing.lex_code =\n%a;\n" output_byte_array tbl.tbl_code;
 
   output_string oc "}\n\n"
@@ -116,8 +132,6 @@ let output_entry some_mem_code ic oc has_refill oci e =
       e.auto_name output_args e.auto_args
 
 (* Main output function *)
-
-exception Table_overflow
 
 let output_lexdef ic oc oci header rh tables entry_points trailer =
   if not !Common.quiet_mode then

--- a/lex/output.ml
+++ b/lex/output.ml
@@ -20,9 +20,10 @@ open Lexgen
 open Compact
 open Common
 
-type c_short_type = Short | Unsigned_short
-
 exception Table_overflow
+
+let check_overflow ~min ~max v =
+  Array.iter (fun n -> if n < min || n > max then raise Table_overflow) v
 
 (* To output an array of short ints, encoded as a string *)
 
@@ -32,28 +33,25 @@ let output_byte oc b =
   output_char oc (Char.chr(48 + (b / 10) mod 10));
   output_char oc (Char.chr(48 + b mod 10))
 
-let output_array c_type oc v =
+let output_array oc v =
   output_string oc "   \"";
   for i = 0 to Array.length v - 1 do
     output_byte oc (v.(i) land 0xFF);
     output_byte oc ((v.(i) asr 8) land 0xFF);
-    let () =
-      let unsigned_number =
-        match c_type with
-        | Short -> v.(i) + 1 lsl 15
-        | Unsigned_short -> v.(i)
-      in
-      if unsigned_number < 0 || unsigned_number >= 1 lsl 16
-      then raise Table_overflow
-    in
     if i land 7 = 7 then output_string oc "\\\n    "
   done;
   output_string oc "\""
 
-let output_array_s = output_array Short
-let output_array_u = output_array Unsigned_short
+let output_array_u oc v =
+  check_overflow ~min:0 ~max: 0xFFFF v;
+  output_array oc v
+
+let output_array_s oc v =
+  check_overflow ~min:(-0x8000) ~max: 0x7FFF v;
+  output_array oc v
 
 let output_byte_array oc v =
+  check_overflow ~min:0 ~max:0xFF v;
   output_string oc "   \"";
   for i = 0 to Array.length v - 1 do
     output_byte oc (v.(i) land 0xFF);

--- a/runtime/lexing.c
+++ b/runtime/lexing.c
@@ -54,8 +54,12 @@ struct lexing_table {
 #define Short(tbl,n) \
   (*((unsigned char *)((tbl) + (n) * 2)) + \
           (*((signed char *)((tbl) + (n) * 2 + 1)) << 8))
+#define UShort(tbl,n) \
+  (*((unsigned char *)((tbl) + (n) * 2)) + \
+          (*((unsigned char *)((tbl) + (n) * 2 + 1)) << 8))
 #else
 #define Short(tbl,n) (((short *)(tbl))[(n)])
+#define UShort(tbl,n) (((unsigned short *)(tbl))[(n)])
 #endif
 
 CAMLprim value caml_lex_engine(value vtbl, value start_state, value vlexbuf)
@@ -176,7 +180,7 @@ CAMLprim value caml_new_lex_engine(value vtbl, value start_state,
     /* Lookup base address or action number for current state */
     base = Short(tbl->lex_base, state);
     if (base < 0) {
-      int pc_off = Short(tbl->lex_base_code, state) ;
+      int pc_off = UShort(tbl->lex_base_code, state) ;
       run_tag(Bp_val(tbl->lex_code) + pc_off, lexbuf->lex_mem);
       /*      fprintf(stderr,"Perform: %d\n",-base-1) ; */
       return Val_int(-base-1);
@@ -184,7 +188,7 @@ CAMLprim value caml_new_lex_engine(value vtbl, value start_state,
     /* See if it's a backtrack point */
     backtrk = Short(tbl->lex_backtrk, state);
     if (backtrk >= 0) {
-      int pc_off =  Short(tbl->lex_backtrk_code, state);
+      int pc_off =  UShort(tbl->lex_backtrk_code, state);
       run_tag(Bp_val(tbl->lex_code) + pc_off, lexbuf->lex_mem);
       lexbuf->lex_last_pos = lexbuf->lex_curr_pos;
       lexbuf->lex_last_action = Val_int(backtrk);
@@ -218,12 +222,12 @@ CAMLprim value caml_new_lex_engine(value vtbl, value start_state,
       }
     }else{
       /* If some transition, get and perform memory moves */
-      int base_code = Short(tbl->lex_base_code, pstate) ;
+      int base_code = UShort(tbl->lex_base_code, pstate) ;
       int pc_off ;
       if (Short(tbl->lex_check_code, base_code + c) == pstate)
-        pc_off = Short(tbl->lex_trans_code, base_code + c) ;
+        pc_off = UShort(tbl->lex_trans_code, base_code + c) ;
       else
-        pc_off = Short(tbl->lex_default_code, pstate) ;
+        pc_off = UShort(tbl->lex_default_code, pstate) ;
       if (pc_off > 0)
         run_mem(Bp_val(tbl->lex_code) + pc_off, lexbuf->lex_mem,
                 lexbuf->lex_curr_pos) ;

